allensdk/brain_observatory/behavior/running_processing.py

```diff
@@ -14,40 +16,378 @@ def calc_deriv(x, time):
+                will reach prior to "wrapping". This should
+                theoretically be 5V (after crossing 5V goes to 0V, or
+                vice versa). In practice the encoder does not always
+                reach this value before wrapping, which can cause
+                transient spikes in speed at the voltage "wraps".
+        The raw data are provided so that the user may compute their
+        own speed from source, if desired.
+
+    Notes
+    -----
+    Though the angular change is available in the raw data
+    (key="dx"), this method recomputes the angular change from the
+    voltage signal (key="vsig") due to very specific, low-level
+    artifacts in the data caused by the encoder. See method
+    docstrings for more detailed information. The raw data is
+    included in the final output in case the end user wants to apply
+    their own corrections and compute running speed from the raw
+    source.
+    """
     v_sig = data["items"]["behavior"]["encoders"][0]["vsig"]
     v_in = data["items"]["behavior"]["encoders"][0]["vin"]
+
     if len(v_in) > len(time) + 1:
-        error_string = "length of v_in ({}) cannot be longer than length of time ({}) + 1, they are off by {}".format(
+        error_string = ("length of v_in ({}) cannot be longer than length of "
+                        "time ({}) + 1, they are off by {}").format(
             len(v_in),
             len(time),
             abs(len(v_in) - len(time))
         )
         raise ValueError(error_string)
     if len(v_in) == len(time) + 1:
         warnings.warn(
-            'time array is 1 value shorter than encoder array. Last encoder value removed\n', stacklevel=1)
-    # remove big, single frame spikes in encoder values
-    dx = scipy.signal.medfilt(dx_raw[:len(time)], kernel_size=5)
-    dx = np.cumsum(dx)  # wheel rotations
-    speed = calc_deriv(dx, time)  # speed in degrees/s
-    speed = deg_to_dist(speed)
+            "Time array is 1 value shorter than encoder array. Last encoder "
+            "value removed\n", stacklevel=1)
+
+    # dx = 'd_theta' = angular change
+    # There are some issues with angular change in the raw data so we
+    # recompute this value
+    dx_raw = data["items"]["behavior"]["encoders"][0]["dx"]
+    # Identify "wraps" in the voltage signal that need to be unwrapped
+    # This is where the encoder switches from 0V to 5V or vice versa
+    pos_wraps, neg_wraps = _identify_wraps(
+        v_sig, min_threshold=1.5, max_threshold=3.5)
+    # Unwrap the voltage signal and apply correction for transient spikes
+    unwrapped_vsig = _unwrap_voltage_signal(
+        v_sig, pos_wraps, neg_wraps, max_threshold=5.1, max_diff=1.0)
+    angular_change_point = _angular_change(unwrapped_vsig, v_in)
+    angular_change = np.nancumsum(angular_change_point)
+    # Add the nans back in (get turned to 0 in nancumsum)
+    ang_change_nans = np.isnan(angular_change_point).nonzero()[0]
+    angular_change[ang_change_nans] = np.nan
```

Comment:
again, can utilze numpy boolean to eliminate a line, 2 operations. Just makes it more clear:
```
In [20]: a = np.array([1.0, 1.2, 1.1, np.NaN, 2.2, 1.05, 1.2, 1.01, 1.01])

In [21]: b = np.nancumsum(a)

In [22]: b
Out[22]: array([1.  , 2.2 , 3.3 , 3.3 , 5.5 , 6.55, 7.75, 8.76, 9.77])

In [23]: a.shape
Out[23]: (9,)

In [24]: b.shape
Out[24]: (9,)

In [25]: b[np.isnan(a)] = np.nan

In [26]: b
Out[26]: array([1.  , 2.2 , 3.3 ,  nan, 5.5 , 6.55, 7.75, 8.76, 9.77])

```